### PR TITLE
IDEMPIERE-5356 - Add Printing Support to ProcessInfoDialog

### PR DESCRIPTION
**Pull request by PeterTakacs300**
JIRA ticket: https://idempiere.atlassian.net/browse/IDEMPIERE-5356

### CODE REVIEW

**Comment by PeterTakacs300** [comment]
Hi @hengsin , I'm not sure how you ment _"printing should depends solely on whether the linked document is printable"_. For now, we tried similar approach as in ReportCtl.start method.

**Comment by hengsin** [comment]
I means whether the linked document is with a document Report Engine Type. In other words, instead of hard coding the process info id, you check whether the ad_table_id of the process log record is associated with a document Report Engine Type (Shipment, Invoice, Order, etc). Another alternative is to add a report engine type field to the process log and use that to decide whether printing is available.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java [hunk 1]
```diff
@@ -216,6 +245,64 @@ public void onEvent(Event event) throws Exception {
...
+	/**
+	 * Is Printable
+	 * @param pi
+	 * @return
+	 */
+	public boolean isPrintable(ProcessInfo pi) {
+		for(ProcessInfoLog log : m_logs)
+			for(int tableID : ReportEngine.getDocTableIDs())
+				if(log.getAD_Table_ID() == tableID) {
+					reportEngineType = ReportEngine.getReportEngineType(tableID);
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java:302]: should check reportEngineType is valie (i.e not -1) here.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java [hunk 2]
```diff
@@ -216,6 +245,64 @@ public void onEvent(Event event) throws Exception {
...
+		// Loop through all items
+		List<File> pdfList = new ArrayList<File>();
+		for (int i = 0; i < m_logs.length; i++)
+		{
+			int RecordID = m_logs[i].getRecord_ID();
+			ReportEngine re = null;
+			
+			re = ReportEngine.get (Env.getCtx(), reportEngineType, RecordID);
+			
+			pdfList.add(re.getPDF());
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java:267]: This code block only work if all m_logs record is of same ad_table_id and with record_id > 0. I don't think it is a good idea to assume that's always the case.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java [hunk 3]
```diff
@@ -51,9 +64,17 @@ public class ProcessInfoDialog extends Window implements EventListener<Event> {
 
 	private static final String MESSAGE_PANEL_STYLE = "text-align:left; word-break: break-all; overflow: auto; max-height: 250pt; min-width: 230pt; max-width: 450pt;";
 
+	/**	Logger			*/
+	private static final CLogger log = CLogger.getCLogger(WGenForm.class);
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java:68]: WGenForm.class is wrong here.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java [hunk 4]
```diff
@@ -216,6 +245,64 @@ public void onEvent(Event event) throws Exception {
 		if (event.getTarget() == btnOk) {
 			this.detach();
 		}
+		if(event.getTarget() == btnPrint) {
+			Clients.showBusy("Processing...");
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java:249]: not translated.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java [hunk 5]
```diff
@@ -216,6 +245,64 @@ public void onEvent(Event event) throws Exception {
...
+			int RecordID = m_logs[i].getRecord_ID();
+			ReportEngine re = null;
+			
+			re = ReportEngine.get (Env.getCtx(), reportEngineType, RecordID);
+			
+			pdfList.add(re.getPDF());				
+		}
+		if (pdfList.size() > 1) {
+			try {
+				File outFile = File.createTempFile(printTmpFileName, ".pdf");
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java:271]: printTmpFileName is not being set any where ...

File: org.adempiere.base/src/org/compiere/print/ReportEngine.java [hunk 6]
```diff
@@ -2402,6 +2402,28 @@ public void setIsReplaceTabContent(boolean m_isReplaceTabContent) {
 		this.m_isReplaceTabContent = m_isReplaceTabContent;
 	}
 
+	/**
+	 * Get DOC_TABLE_ID array
+	 * @return
+	 */
+	public static int[] getDocTableIDs() {
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/print/ReportEngine.java:2409]: this public method is not needed

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java (hunk 1 above)
Comment (hengsin) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java:302]: this loop of ReportEngine.getDocTableIDs() is not good, it should be simplify to just `if (ReportEngine.getReportEngineType(tableID) >= 0) return true;`

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java [hunk 7]
```diff
@@ -216,6 +242,63 @@ public void onEvent(Event event) throws Exception {
...
+	public void onPrint() {
+		// Loop through all items
+		List<File> pdfList = new ArrayList<File>();
+		for (int i = 0; i < m_logs.length; i++)
+		{
+			int recordID = m_logs[i].getRecord_ID();
+			ReportEngine re = null;
+			
+			if(recordID > 0) {
+				re = ReportEngine.get (Env.getCtx(), ReportEngine.getReportEngineType(m_logs[i].getAD_Table_ID()), recordID);
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java:263]: the safe thing to do here is `int reportEngineType = ReportEngine.getReportEngineType(m_logs[i].getAD_Table_ID()); if (reportEngineType >= 0) { ...}`

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java [hunk 8]
```diff
@@ -216,6 +242,64 @@ public void onEvent(Event event) throws Exception {
...
+		if (pdfList.size() > 1) {
+			try {
+				File outFile = File.createTempFile(getTitle(), ".pdf");					
+				AEnv.mergePdf(pdfList, outFile);
+
+				Clients.clearBusy();
+				Window win = new SimplePDFViewer(getTitle(), new FileInputStream(outFile));
+				SessionManager.getAppDesktop().showWindow(win, "center");
+			} catch (Exception e) {
+				log.log(Level.SEVERE, e.getLocalizedMessage(), e);
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java:277]: Better just rethrow as runtime exception instead. Otherwise it is silent error to user.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java [hunk 9]
```diff
@@ -216,6 +242,64 @@ public void onEvent(Event event) throws Exception {
...
+			}
+		} else if (pdfList.size() > 0) {
+			Clients.clearBusy();
+			try {
+				Window win = new SimplePDFViewer(getTitle(), new FileInputStream(pdfList.get(0)));
+				SessionManager.getAppDesktop().showWindow(win, "center");
+			} catch (Exception e)
+			{
+				log.log(Level.SEVERE, e.getLocalizedMessage(), e);
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java:286]: Better just rethrow as runtime exception instead. Otherwise it is silent error to user.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java [hunk 10]
```diff
@@ -216,6 +242,64 @@ public void onEvent(Event event) throws Exception {
 		if (event.getTarget() == btnOk) {
 			this.detach();
 		}
+		if(event.getTarget() == btnPrint) {
+			Clients.showBusy(Msg.getMsg(Env.getCtx(), "Processing"));
+			Clients.response(new AuEcho(ProcessInfoDialog.this, "onPrint", null));
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java:247]: the alternative is to use `Executions.schedule(this.getDesktop(), e -> onPrint(), new Event("onPrint"));`. The benefit is you can declare onPrint as private and didn't add a "fake" public method to ProcessInfoDialog (fake as in although it is public but it is not means to be call directly by other class).

**Review by hengsin — CHANGES_REQUESTED** [review]

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java [hunk 11]
```diff
@@ -216,6 +237,64 @@ public void onEvent(Event event) throws Exception {
...
+		if (pdfList.size() > 1) {
+			try {
+				File outFile = FileUtil.createTempFile(getTitle(), ".pdf");					
+				AEnv.mergePdf(pdfList, outFile);
+
+				Clients.clearBusy();
+				Window win = new SimplePDFViewer(getTitle(), new FileInputStream(outFile));
+				SessionManager.getAppDesktop().showWindow(win, "center");
+			} catch (Exception e) {
+				throw new RuntimeException(e.getLocalizedMessage());
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java:272]: best practise is to use throw new RuntimeException(e.getLocalizedMessage(), e); instead so the exception stack trace is not lost. same goes for the throw below.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java [hunk 12]
```diff
@@ -216,6 +237,64 @@ public void onEvent(Event event) throws Exception {
...
+				re = ReportEngine.get (Env.getCtx(), reportEngineType, recordID);
+				pdfList.add(re.getPDF());
+			}
+		}
+		if (pdfList.size() > 1) {
+			try {
+				File outFile = FileUtil.createTempFile(getTitle(), ".pdf");					
+				AEnv.mergePdf(pdfList, outFile);
+
+				Clients.clearBusy();
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java:268]: this should goes before "try{", same as the Clients.clearBusy() call below.

**Review by hengsin — CHANGES_REQUESTED** [review]

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java [hunk 13]
```diff
@@ -216,6 +237,64 @@ public void onEvent(Event event) throws Exception {
...
+			Clients.showBusy(Msg.getMsg(Env.getCtx(), "Processing"));
+			Executions.schedule(this.getDesktop(), e -> onPrint(), new Event("onPrint"));
+		}
+	}
+	
+	/**
+	 * On Print
+	 */
+	private void onPrint() {
+		// Loop through all items
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java:250]: hmm ... sorry for the mistake in previous review, Clients.clearBusy(); should be the first line in this method, not the 2 place below. No matter what happens below, we need to make sure the busy mask is clear.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java [hunk 14]
```diff
@@ -216,6 +237,64 @@ public void onEvent(Event event) throws Exception {
...
+				throw new RuntimeException(e.getLocalizedMessage(), e);
+			}
+		}
+		this.detach();
+	}
+	
+	/**
+	 * Is Printable
+	 * @param pi
+	 * @return
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/ProcessInfoDialog.java:290]: this is minor but @return is empty here.

**Review by hengsin — CHANGES_REQUESTED** [review]
